core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -182,20 +188,14 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
                 // if its a conflict failure, and we already executed the request on a primary (and we execute it
                 // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
                 // then just use the response we got from the successful execution
-                if (item.getPrimaryResponse() == null || isConflictException(failure) == false) {
-                    item.setPrimaryResponse(new BulkItemResponse(item.id(), docWriteRequest.opType(),
+                if (replicaRequest.getPrimaryResponse() == null || isConflictException(failure) == false) {
```

Comment:
This is a mess - there is no guarantee from the coe that the replica request wasn't just replaced when we did `request.items()[requestIndex] = replicaRequest;` . In practice I think it's OK, but I don't like it. Also, now that we have a proper primary relocation (where we don't shut down the shards half way operations), I don't think it's needed. Can you open another follow up issue so we can evaluate carefully?